docs/markdown/Writing Plugins/tutorials/01-create-new-goal-with-a-custom-target.md

```diff
@@ -0,0 +1,335 @@
+# Introduction
+
+In this tutorial, you'll learn the basics needed to get started writing a plugin. You will create a new goal, `project-version`, which will tell you the version (retrieved from the `VERSION` text file) of a particular project in your monorepository. You will learn how to create a new custom target to refer to the `VERSION` file, how to author a new goal, and, most importantly, how to connect rules and targets. You can follow along this tutorial in your own repository; you only need to be on a recent version of Pants and have a `VERSION` file containing a version string e.g. `1.2.3`.
+
+## Registering a plugin
+
+We'll be writing an [in-repo plugin](https://www.pantsbuild.org/docs/plugins-overview#in-repo-plugins), and expect you to have the `pants-plugins/project_version` directory with the `pants.toml` configuration file extended:
+
+```toml pants.toml
+pythonpath = ["%(buildroot)s/pants-plugins"]
+
+backend_packages = [
+  "pants.backend.python",
+   ...
+  "project_version",
+]
+```
+
+## Creating a new target
+
+Once you have become familiar with the [core concepts of Targets and Fields](https://www.pantsbuild.org/docs/target-api-concepts), you are ready to [create an own custom target](https://www.pantsbuild.org/docs/target-api-new-targets) that will represent the `VERSION` file:
+
+```python pants-plugins/project_version/targets.py
+from pants.engine.target import COMMON_TARGET_FIELDS, SingleSourceField, Target
+
+
+class ProjectVersionTarget(Target):
+    alias = "version_file"
+    core_fields = (*COMMON_TARGET_FIELDS, SingleSourceField)
+    help = "A project version target representing the VERSION file."
+```
+
+Our target has some common target fields such as `tags` and `description` available via the `COMMON_TARGET_FIELDS`; including those fields in your targets may be convenient if you decide to use tags and provide a description later. In addition, it also has the [`source` field](https://www.pantsbuild.org/docs/rules-api-and-target-api#sourcesfield) which will be used to provide path to the project's `VERSION` file. 
+
+We could [add a custom field](https://www.pantsbuild.org/docs/target-api-new-fields) to provide a file path, however, when using the `source` field, you get a few features for free such as setting the `default` value and `expected_file_extensions`. Furthermore, with the `source` field, thanks to the [`unmatched_build_file_globs`](https://www.pantsbuild.org/docs/reference-global#unmatched_build_file_globs) option, you won't need to provide custom logic to handle errors when a path glob doesn't expand to any files in your repository.
+
+You only need to register the target to start using it: 
+
+```python pants-plugins/project_version/register.py
+from project_version.targets import ProjectVersionTarget
+
+
+def target_types():
```

Comment:
It would be a huge +1 from me if our docs were type-hinted as much as possible. Makes discovery soooo much easier:

I -think- it's this? At least, that's what I use, but I'm not sure if it's supposed to be `Iterable[Target]`

` -> Iterable[type[Target]]:`

Reply:
yea, `type[Target]` as it's not instances of them... but the types ;)